jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/RegistryServiceTest.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.build.service.docker;
+
+import org.eclipse.jkube.kit.build.service.docker.access.DockerAccess;
+import org.eclipse.jkube.kit.common.KitLogger;
+import org.eclipse.jkube.kit.common.RegistryConfig;
+import org.eclipse.jkube.kit.config.image.ImageConfiguration;
+import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
+import org.eclipse.jkube.kit.config.image.build.ImagePullPolicy;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Answers;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RegistryServiceTest {
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private KitLogger mockedLogger;
+
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private DockerAccess mockedDockerAccess;
+
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private RegistryConfig mockedRegistryConfig;
+
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private ImagePullManager mockedImagePullManager;
+
+  private RegistryService registryService;
+
+  private ImageConfiguration imageConfiguration;
+
+  @Before
+  public void setUp() {
+    imageConfiguration = ImageConfiguration.builder()
+        .name("foo/bar:0.0.1")
+        .build(BuildConfiguration.builder()
+            .from("basefoo/basebar:0.1.0")
+            .tags(Arrays.asList("latest", "0.0.1-slim"))
+            .build())
+        .build();
+    registryService = new RegistryService(mockedDockerAccess, mockedLogger);
+    when(mockedRegistryConfig.getRegistry()).thenReturn("example.com");
+  }
+
+  @Test
+  public void pushImages_whenValidImageConfigurationProvidedWithSkipTag_shouldPushImages() throws IOException {
+    // When
+    registryService.pushImage(imageConfiguration, 1, mockedRegistryConfig, true);
+
+    // Then
+    verify(mockedDockerAccess, times(1))
+        .pushImage(eq("foo/bar:0.0.1"), any(), eq("example.com"), anyInt());
+    verify(mockedDockerAccess, times(0))
+        .pushImage(eq("foo/bar:latest"), any(), eq("example.com"), anyInt());
+    verify(mockedDockerAccess, times(0))
+        .pushImage(eq("foo/bar:0.0.1-slim"), any(), eq("example.com"), anyInt());
```

Comment:
I'm not following these test scenarios.

Why do we have 3 verifications?

methods should also be renamed to + `pushImage_...`

Reply:
these three verifications are verifying all three tags are pushed. One specified in name and two specified in `<tags>`

Reply:
OK, maybe a clearer expectation name can help

`pushImages_whenValidImageConfigurationProvidedWithSkipTag_shouldPushApplicableTags`

or

`pushImages_whenValidImageConfigurationProvidedWithSkipTag_shouldNotPushAdditionalTags`


And the opposite for the other test.

Reply:
The other 2 verifications could be replaced by
```
verify(mockedDockerAccess, times(1))
    .pushImage(eq("foo/bar:0.0.1"), any(), eq("example.com"), anyInt());
verify(mockedDockerAccess, times(1))
        .pushImage(any(), any(), any(), anyInt());
```

Which makes the test more reliable (you verify that nothing else, despite the name, is pushed)